### PR TITLE
Using Versionless rids instead of custom OSes

### LINKED ISSUE #32451: Use RID-layout for OS configurations in libraries
We should switch to the well defined RID system away from our current defined OS configurations. This allows us to share the OS configuration matrix with installer and have just one central point where this is defined.

See https://github.com/dotnet/runtime/blob/97044930e2248d0d529d974e867e551fcffc7abf/src/installer/Directory.Build.props#L227-L308 for how this is currently implemented in installers.

cc @Anipik @ericstj

### PR DESCRIPTION
**Pull request by Anipik**
Fixes https://github.com/dotnet/runtime/issues/32451

### CODE REVIEW

File: src/libraries/Directory.Build.props [hunk 1]
```diff
@@ -23,12 +23,12 @@
     <IsSourceProject>$([System.Text.RegularExpressions.Regex]::IsMatch($(MSBuildProjectDirectory), 'src%24'))</IsSourceProject>
     <LanguageTargets Condition="'$(MSBuildProjectExtension)' == '.depproj'">$(RepositoryEngineeringDir)depProj.common.targets</LanguageTargets>
     <RuntimePropsFile>$([MSBuild]::NormalizePath('$(ArtifactsDir)', 'tmp', '$(Configuration)', 'RuntimeOS.props'))</RuntimePropsFile>
-    <RuntimeGraph>$(LibrariesProjectRoot)\OSGroups.json</RuntimeGraph>
+    <RuntimeGraph>$(LibrariesProjectRoot)\pkg\Microsoft.NETCore.Platforms\runtime.json</RuntimeGraph>
```
Comment (ericstj) [inline src/libraries/Directory.Build.props]: Can you use https://github.com/dotnet/runtime/blob/33a6ed57e354f270d86ac4b337c07691298093f2/Directory.Build.props#L68

**Review by ericstj — APPROVED** [review]
did a scan of libraries + plumbing and it seemed ok.  I want to make sure we have folks from coreclr and installer review as well.  This one may warrant a manual official build.

File: eng/pipelines/common/xplat-setup.yml [hunk 2]
```diff
@@ -69,31 +69,31 @@ jobs:
 
     ${{ if eq(parameters.jobParameters.pool, '') }}:
       pool:
-        # Public Linux Build Pool
```
Comment (jkotas) [inline eng/pipelines/common/xplat-setup.yml]: The lower casing of Linux in comments like this one should be reverted.

**Comment by jkotas** [comment]
The abbreviated and all lower identifiers look odd when everything else in .NET is non-abbreviated and CamelCased.

**Comment by ericstj** [comment]
@jkotas the same could be said for TargetFrameworks. These come from the same source.

File: eng/native/init-distro-rid.sh [hunk 3]
```diff
@@ -150,18 +150,18 @@ initDistroRidGlobal()
         export __PortableBuild
         local distroRid=""
 
-        # Check for musl-based distros (e.g Alpine Linux, Void Linux).
+        # Check for musl-based distros (e.g Alpine linux, Void linux).
```
Comment (am11) [inline eng/native/init-distro-rid.sh]: nit: comments can stay in upper case `Linux`.

File: eng/pipelines/common/platform-matrix.yml [hunk 4]
```diff
@@ -18,7 +18,7 @@ parameters:
   # for the given platform and helixQueueGroup.
   helixQueuesTemplate: ''
   stagedBuild: false
-  # When set to false, suppresses reuse of OSX managed build artifacts (for pipelines without an OSX obj)
+  # When set to false, suppresses reuse of osx managed build artifacts (for pipelines without an osx obj)
```
Comment (am11) [inline eng/pipelines/common/platform-matrix.yml]:
```suggestion
  # When set to false, suppresses reuse of macOS managed build artifacts (for pipelines without an macOS obj)
```

File: eng/pipelines/common/platform-matrix.yml [hunk 5]
```diff
@@ -27,14 +27,14 @@ parameters:
 
 jobs:
 
-# Linux arm
+# linux arm
```
Comment (am11) [inline eng/pipelines/common/platform-matrix.yml]:
```suggestion
# Linux arm
```

File: eng/native/init-distro-rid.sh (hunk 3 above)
Comment (am11) [reply eng/native/init-distro-rid.sh]: ah, just saw @jkotas's comment above. please ignore :)

File: src/libraries/Directory.Build.props [hunk 6]
```diff
@@ -301,7 +258,7 @@
     <!-- Helix properties -->
     <OSPlatformConfig>$(TargetOS).$(Platform).$(Configuration)</OSPlatformConfig>
     <AnyOSPlatformConfig>AnyOS.AnyCPU.$(Configuration)</AnyOSPlatformConfig>
```
Comment (jkotas) [inline src/libraries/Directory.Build.props]: Lower case AnyOS and AnyCPU as well?

File: eng/common/templates/job/job.yml [hunk 7]
```diff
@@ -106,7 +106,7 @@ jobs:
         env:
           TeamName: $(_TeamName)
         continueOnError: ${{ parameters.continueOnError }}
-        condition: and(succeeded(), in(variables['_SignType'], 'real', 'test'), eq(variables['Agent.Os'], 'Windows_NT'))
+        condition: and(succeeded(), in(variables['_SignType'], 'real', 'test'), eq(variables['Agent.Os'], 'win'))
```
Comment (ViktorHofer) [inline eng/common/templates/job/job.yml]: Has the `Agent.Os` value actually changed? Isn't this something coming from AzDO?

**Review by ViktorHofer — CHANGES_REQUESTED** [review]
Please revert changes in eng/common

File: eng/pipelines/common/platform-matrix.yml [hunk 8]
```diff
@@ -47,18 +47,18 @@ jobs:
         ${{ if eq(parameters.passPlatforms, true) }}:
           platforms: ${{ parameters.platforms }}
         helixQueueGroup: ${{ parameters.helixQueueGroup }}
-        managedTestBuildOsGroup: Linux
+        managedTestBuildOsGroup: linux
```
Comment (ViktorHofer) [inline eng/pipelines/common/platform-matrix.yml]: @Anipik did we miss these when you renamed OSGroup to BuildOS?

File: src/libraries/Directory.Build.props [hunk 9]
```diff
@@ -52,59 +52,16 @@
...
+  <PropertyGroup>
+    <TargetsFreeBSD Condition="'$(TargetFrameworkSuffix)' == 'freebsd'">true</TargetsFreeBSD>
+    <TargetsLinux Condition="'$(TargetFrameworkSuffix)' == 'linux'">true</TargetsLinux>
+    <TargetsNetBSD Condition="'$(TargetFrameworkSuffix)' == 'netbsd'">true</TargetsNetBSD>
+    <TargetsOSX Condition="'$(TargetFrameworkSuffix)' == 'osx'">true</TargetsOSX>
+    <TargetsWindows Condition="'$(TargetFrameworkSuffix)' == 'win'">true</TargetsWindows>
+    <TargetsUnix Condition="'$(TargetFrameworkSuffix)' == 'unix' or '$(TargetsLinux)' == 'true' or '$(TargetsNetBSD)' == 'true' or '$(TargetsOSX)' == 'true' or '$(TargetsFreeBSD)' == 'true'">true</TargetsUnix>
+    <TargetsAnyOS Condition="'$(TargetFrameworkSuffix)' == ''">true</TargetsAnyOS>  
+    <PackageTargetRuntime Condition="'$(TargetFrameworkSuffix)' != ''">$(TargetFrameworkSuffix)</PackageTargetRuntime>
+  </PropertyGroup>
```
Comment (ViktorHofer) [inline src/libraries/Directory.Build.props]: TargetsWebAssembly?

File: eng/pipelines/common/platform-matrix.yml (hunk 8 above)
Comment (Anipik) [reply eng/pipelines/common/platform-matrix.yml]: I didnt change the .yml scripts variable names in the previous PR. i Wanted to keep that little bit separate as other change was getting crowded.

File: eng/native/configureplatform.cmake [hunk 10]
```diff
@@ -275,7 +275,7 @@ if (CLR_CMAKE_TARGET_OS STREQUAL CLR_CMAKE_HOST_OS)
         endif()
     endif()
 else()
-    if(NOT (CLR_CMAKE_HOST_OS STREQUAL Windows_NT))
+    if(NOT (CLR_CMAKE_HOST_OS STREQUAL win))
```
Comment (akoeplinger) [inline eng/native/configureplatform.cmake]: All the other `CLR_CMAKE_HOST_OS` values on other platform are initialized to longer names that start with upper case so this makes it inconsistent.

You could change it to `Windows` on L117 instead.

**Review by akoeplinger — CHANGES_REQUESTED** [review]

File: src/installer/pkg/projects/Microsoft.NETCore.DotNetHost/runtime.osx.Microsoft.NETCore.DotNetHost.props [hunk 11]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (jkotas) [inline src/installer/pkg/projects/Microsoft.NETCore.DotNetHost/runtime.osx.Microsoft.NETCore.DotNetHost.props]: Delete the file with upper-cased name?

Comment (Anipik) [reply src/installer/pkg/projects/Microsoft.NETCore.DotNetHost/runtime.osx.Microsoft.NETCore.DotNetHost.props]: I double checked the upper case files are not present, for some reason these files are not being shown as rename by github

Comment (jkotas) [reply src/installer/pkg/projects/Microsoft.NETCore.DotNetHost/runtime.osx.Microsoft.NETCore.DotNetHost.props]: I guess you are on Windows and git got confused by the Windows case-insensitive file system.

You need to go to Unix where you can see the files with different casing and clean it up there: 
```
jkotas@JKOTAS5:~/runtime/src/installer/pkg/projects/Microsoft.NETCore.DotNetHost$ ls -la
total 36
drwxrwxrwx 1 jkotas jkotas 4096 Mar  9 16:29 .
drwxrwxrwx 1 jkotas jkotas 4096 Mar  9 16:29 ..
-rw-rw-rw- 1 jkotas jkotas 2654 Mar  9 16:24 host.wxs
-rw-rw-rw- 1 jkotas jkotas 2039 Mar  9 16:24 Microsoft.NETCore.DotNetHost.pkgproj
-rw-rw-rw- 1 jkotas jkotas  439 Mar  9 16:29 runtime.freebsd.Microsoft.NETCore.DotNetHost.props
-rw-rw-rw- 1 jkotas jkotas  439 Mar  9 16:24 runtime.FreeBSD.Microsoft.NETCore.DotNetHost.props
-rw-rw-rw- 1 jkotas jkotas  439 Mar  9 16:29 runtime.osx.Microsoft.NETCore.DotNetHost.props
-rw-rw-rw- 1 jkotas jkotas  439 Mar  9 16:24 runtime.OSX.Microsoft.NETCore.DotNetHost.props
-rw-rw-rw- 1 jkotas jkotas  439 Mar  9 16:29 runtime.unix.Microsoft.NETCore.DotNetHost.props
-rw-rw-rw- 1 jkotas jkotas  439 Mar  9 16:24 runtime.Unix.Microsoft.NETCore.DotNetHost.props
-rw-rw-rw- 1 jkotas jkotas  443 Mar  9 16:29 runtime.win.Microsoft.NETCore.DotNetHost.props
```

Comment (Anipik) [reply src/installer/pkg/projects/Microsoft.NETCore.DotNetHost/runtime.osx.Microsoft.NETCore.DotNetHost.props]: okay i will do that.

**Comment by dagood** [comment]
> This one may warrant a manual official build.

I agree, please include a link here to a validation mock official build here once you have one, to keep track of it.

Note there have been changes to the sharedfx SDK already to allow for this change, I think: https://github.com/dotnet/arcade/pull/4911 removes the OSGroup dependency. Otherwise, nothing jumps out to me as wrong (except the existing comments here).

File: eng/common/templates/job/execute-sdl.yml [hunk 12]
```diff
@@ -11,7 +11,7 @@ parameters:
                                                                # Usage:
```
Comment (safern) [inline eng/common/templates/job/execute-sdl.yml]: All the changes under /eng/common/* should be reverted, these are files that live in arcade.

File: eng/common/templates/job/job.yml (hunk 7 above)
Comment (safern) [reply eng/common/templates/job/job.yml]: Yeah, this is wrong. Anyway, this file shouldn't be changed as it is an arcade file

File: eng/pipelines/libraries/outerloop.yml [hunk 13]
```diff
@@ -72,12 +72,12 @@ jobs:
         buildConfig: Debug
         platforms:
         - ${{ if eq(variables['includeWindowsOuterloop'], true) }}:
-          - Windows_NT_x64
+          - win_x64
         - ${{ if eq(variables['includeLinuxOuterloop'], true) }}:
```
Comment (NikolaMilosavljevic) [inline eng/pipelines/libraries/outerloop.yml]: Would this need to be lowercase to match other changes - 'includelinuxOuterloop'?

Curious why only Linux-one was being changed to lowercase and not Windows and Osx?

File: eng/run-test.sh [hunk 14]
```diff
@@ -285,7 +285,7 @@ if [ $RunTestSequential -eq 1 ]
 then
     maxProcesses=1;
 else
-    if [ `uname` = "NetBSD" ] || [ `uname` = "FreeBSD" ]; then
+    if [ `uname` = "NetBS" ] || [ `uname` = "FreeBSD" ]; then
```
Comment (NikolaMilosavljevic) [inline eng/run-test.sh]: All instances of 'NetBS' seem odd - is this intentional?

**Review by NikolaMilosavljevic — CHANGES_REQUESTED** [review]

**Comment by jkotas** [comment]
> > The abbreviated and all lower identifiers look odd when everything else in .NET is non-abbreviated and CamelCased.

> @jkotas the same could be said for TargetFrameworks. These come from the same source.

FWIW, the current proposal is to use proper casing for the solitary OS names: https://github.com/dotnet/designs/blob/4284705c22ed180753e0efef2ca72698de8e27b6/accepted/2020/net5/net5.md#mapping-to-properties

**Comment by ericstj** [comment]
> FWIW, the current proposal is to use proper casing for the solitary OS names in .NET 5: https://github.com/dotnet/designs/blob/4284705c22ed180753e0efef2ca72698de8e27b6/accepted/2020/net5/net5.md#mapping-to-properties

Yep, but still lower case and abbreviated in TargetFramework. We'll follow whatever the product does here if there is a better path.  For now the best path is to follow NuGet precedent with lower-case abbreviated TargetFramework values and RIDs.

File: eng/pipelines/coreclr/templates/run-performance-job.yml [hunk 15]
```diff
@@ -90,7 +90,7 @@ jobs:
       displayName: Build scenario tools
       env:
         PERFLAB_TARGET_FRAMEWORKS: netcoreapp3.1
-      condition: and(succeeded(), eq(variables['Agent.Os'], 'Windows_NT'))
+      condition: and(succeeded(), eq(variables['Agent.Os'], 'win'))
```
Comment (safern) [inline eng/pipelines/coreclr/templates/run-performance-job.yml]: Whatever uses `Agent.OS` shouldn't be changed. Those values are set by Azure DevOps and they return full OS names.

**Comment by Anipik** [comment]
@ericstj @ViktorHofer @safern can you take a look at this. Most of the failures are resolved here and i have a pr in jitutils to fix the formatting failure.

File: src/libraries/Directory.Build.props (hunk 6 above)
Comment (ericstj) [reply src/libraries/Directory.Build.props]: AnyCPU comes from VS and this is the casing they use.
`AnyOS` is our invention, and perhaps it should be eliminated.  There is a representation in the RID graph:
https://github.com/dotnet/runtime/blob/1124c1ac95bf1b74b4a030fc21ef0b9f58e33b07/src/libraries/pkg/Microsoft.NETCore.Platforms/runtime.json#L370

File: src/libraries/pkg/Microsoft.NETCore.Platforms/runtimeGroups.props [hunk 16]
```diff
@@ -81,6 +81,10 @@
       <Versions>1;2</Versions>
     </RuntimeGroup>
 
+    <RuntimeGroup Include="WebAssembly">
```
Comment (ericstj) [inline src/libraries/pkg/Microsoft.NETCore.Platforms/runtimeGroups.props]: This is a product change, did you mean to include this?  I thought this was still under discussion: https://github.com/dotnet/runtime/issues/33327

File: eng/native/configureplatform.cmake [hunk 17]
```diff
@@ -86,6 +87,7 @@ if(CLR_CMAKE_HOST_OS STREQUAL Darwin)
     set(CLR_CMAKE_HOST_UNIX_AMD64 1)
     set(CLR_CMAKE_HOST_DARWIN 1)
     set(CMAKE_ASM_COMPILE_OBJECT "${CMAKE_C_COMPILER} <FLAGS> <DEFINES> <INCLUDES> -o <OBJECT> -c <SOURCE>")
+    set(CLR_CMAKE_HOST_OS osx)
```
Comment (jkotas) [inline eng/native/configureplatform.cmake]: What is the convention for `CLR_CMAKE_HOST_OS` ? This line has it lower cased, and the very next line has it upped cased?

Comment (jkotas) [reply eng/native/configureplatform.cmake]: Again, this feels pretty unnatural to use the lower-cased shortened OS names in CMake. The convention in CMake is to use Linux, Windows, etc.

**Comment by jkotas** [comment]
I am sorry, but I still do not understand the benefit of this change. The RID OS names everywhere look plain ugly, we still have the good names in many places and so this is just moving the inconsistency to a different spot. I understand that the installers have to figure out the RID, and this change will make it a bit simpler. How many lines are we going to save by doing this? Are there bad problems with this from the past that this is going to avoid?

**Comment by BruceForstall** [comment]
@Anipik Can you explain in the initial comment above what the PR title actually means, and what its implication is for the various components, including dev workflow? The PR title and associated "Fixed" issue are not sufficiently detailed for me to understand what this is about.

**Comment by BruceForstall** [comment]
Looking through the change, this looks like an extremely disruptive change. There will be many individual scripts (outside the repo) that need to be adjusted. It also seems to turn RID concepts into user-visible concepts, possibly unnecessarily. I always disliked the usage of "Windows_NT" (IMO, it should be "Windows" - we stopped saying "NT" like 20 years ago), and "OSX" should be "macOS". But using "win" everywhere that an OS name is required seems like overkill. Especially since "win" is an English word as well as a magic RID abbreviation meaning the Windows OS. Using "win" instead of "Windows" in non-RID contexts seems wrong. Is this whole change worth the trouble (and inevitable long tail fallout)?

**Comment by ericstj** [comment]
The benefit here is that we change the cross-targeting parameter for the libraries projects to match the actual mechanism that is used to select libraries and our runtime implementations (RID).  This concept and value space is something built into our product, as opposed to the current set of values which is something only that's really just set of strings we use in comparisons with a somewhat organic history.

I agree that this is a pretty huge change touching more than just libraries.  There is some benefit to reducing the number of unique concepts and manual mappings that need to be understood.  I believe the idea here was that after this diff folks could think about a runtime in a single language / value space.  Initially I was just planning on taking this to our libraries infra, but @Anipik generalized to the entire repo in order to minimize the number of "OS" concepts we had floating around.

**Comment by Anipik** [comment]
@jkotalik @BruceForstall @ericstj what is  the final agreement here ?

**Comment by jkotas** [comment]
My take:
- Changes that affects developer work should be synchronized using https://github.com/dotnet/runtime/issues/33821
- For anything that gets renamed to use RIDs, make it follow RIDs 100%: i.e.: Debug\win-x64 instead of win.x64.Debug, linux-arm instead linux_arm, etc.
- Renaming of the build output directories to use RIDs sounds reasonable. I agree with Eric that the .NET SDK uses RID based directory names, and so it sounds reasonable for consistency.
- Can this be broken into multiple changes? E.g. Would it make sense to rename the output directories as the step 1, without changing the other stuff? Or can the renames in the CI config be separate change?
- The non-RID names are used in number of places, and we won't be able to change all of them and achive 100% consistency. I do not think it makes sense to try to change as much as possible to use RID OS names. I think we should rather strike the right ballance, e.g. switch to use RIDs where there is a clear prior art (e.g. where the .NET SDK is using them) and leave the rest alone. For example, I do not think it makes sense to be changing the CMakeFiles to use RID OS names.

**Comment by Anipik** [comment]
@jkotas that looks like a good plan. i will start working on it and try to split in multiple chunks

**Pull request closed without merging**